### PR TITLE
Introduces the `isArtifact` test for symbols

### PR DESCRIPTION
**Pull request by xeno-by**
The test seems to be necessary to tell apart useful synthetic symbols
(such as accessors) and low-level compilation artifacts (such as $outer).

### CODE REVIEW

**Comment by xeno-by** [comment]
review by @paulp or @adriaanm

File: src/reflect/scala/reflect/api/Symbols.scala [hunk 1]
```diff
@@ -67,6 +67,11 @@ trait Symbols extends base.Symbols { self: Universe =>
      */
     def isSynthetic: Boolean
 
+    /** Does this symbol represent a compilation artifact that isn't meant for public use?
+     *  Examples of such artifacts are erasure bridges and $outer fields.
+     */
+    def isArtifact: Boolean
```
Comment (xeno-by) [inline src/reflect/scala/reflect/api/Symbols.scala]: How does this compare with `isSynthetic`?

**Comment by xeno-by** [comment]
Also see http://groups.google.com/group/scala-internals/browse_thread/thread/d04e762127737968

File: src/reflect/scala/reflect/api/Symbols.scala (hunk 1 above)
Comment (paulp) [inline src/reflect/scala/reflect/api/Symbols.scala]: How does this differ from isHidden? As far as I can tell, this is trying to be isHidden with a different name.  Just expose isHidden.

Comment (xeno-by) [inline src/reflect/scala/reflect/api/Symbols.scala]: True, the renaming is intentional. To my liking `isHidden` sounds too generic. WDYT?

Comment (paulp) [inline src/reflect/scala/reflect/api/Symbols.scala]: I agree, but then let's rename HIDDEN and isHidden! We just introduced it a couple weeks ago, it's not like it has some huge installed base of users.  Two names for something is a last resort, not a first.

Comment (paulp) [inline src/reflect/scala/reflect/api/Symbols.scala]: I'd say "isArtifact" is also far too generic; as a compiler hacker maybe it sounds specific, but this isn't even what artifact usually means.  How about isCompilerArtifact.

Comment (paulp) [inline src/reflect/scala/reflect/api/Symbols.scala]: ...although that is not really sufficiently inclusive.  It's more like isImplementationArtifact.

Comment (xeno-by) [inline src/reflect/scala/reflect/api/Symbols.scala]: We can have ARTIFACT and isArtifact in our codebase and expose isImplementationArtifact in the API, okay?

Comment (paulp) [inline src/reflect/scala/reflect/api/Symbols.scala]: I can live with that.

Comment (lrytz) [inline src/reflect/scala/reflect/api/Symbols.scala]: i personally don't think longer names are a problem, so i'd go for `isImplementationArtifact` everywhere.

Comment (xeno-by) [inline src/reflect/scala/reflect/api/Symbols.scala]: I'd say IMPLEMENTATION_ARTIFACT is a bit too much.

Comment (lrytz) [inline src/reflect/scala/reflect/api/Symbols.scala]: maybe..

Comment (paulp) [inline src/reflect/scala/reflect/api/Symbols.scala]: IMPLEMENTATION_ARTIFACT is definitely too much, but the flag doesn't have to match the name exactly.  We should be discouraging direct flag manipulations anyway, internally as well as externally.

But if we are going to eat a really long name, I feel sure we can do better than isImplementationArtifact.

**Comment by xeno-by** [comment]
PLS REBUILD ALL

**Comment by xeno-by** [comment]
PLS REBUID ALL

**Comment by xeno-by** [comment]
hello? PLS REBUILD ALL

**Pull request merged**